rust-runtime/aws-smithy-checksums/src/http.rs

```diff
@@ -0,0 +1,226 @@
+    fn size(&self) -> u64 {
+        let trailer_name_size_in_bytes = self.header_name().as_str().len() as u64;
+        let base64_encoded_checksum_size_in_bytes = base64::encoded_length(Checksum::size(self));
+
+        trailer_name_size_in_bytes
+            // HTTP trailer names and values may be separated by either a single colon or a single
+            // colon and a whitespace. In the AWS Rust SDK, we use a single colon.
+            + ":".len() as u64
+            + base64_encoded_checksum_size_in_bytes
+    }
+}
+
+impl HttpChecksum for Crc32 {
+    fn header_name(&self) -> HeaderName {
+        CRC_32_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.finalize();
+        HeaderValue::from_str(&base64::encode(u32::to_be_bytes(hash)))
+            .expect("will always produce a valid header value from a CRC32 checksum")
+    }
+}
+
+impl HttpChecksum for Crc32c {
+    fn header_name(&self) -> HeaderName {
+        CRC_32_C_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        // If no data was provided to this callback and no CRC was ever calculated, return zero as the checksum.
+        let hash = self.state.unwrap_or_default();
+        HeaderValue::from_str(&base64::encode(u32::to_be_bytes(hash)))
+            .expect("will always produce a valid header value from a CRC32C checksum")
+    }
+}
+
+impl HttpChecksum for Sha1 {
+    fn header_name(&self) -> HeaderName {
+        SHA_1_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        use sha1::Digest;
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.finalize();
+        HeaderValue::from_str(&base64::encode(&hash[..]))
+            .expect("will always produce a valid header value from a SHA-1 checksum")
+    }
+}
+
+impl HttpChecksum for Sha256 {
+    fn header_name(&self) -> HeaderName {
+        SHA_256_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        use sha2::Digest;
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.finalize();
+        HeaderValue::from_str(&base64::encode(&hash[..]))
+            .expect("will always produce a valid header value from a SHA-256 checksum")
+    }
+}
+
+impl HttpChecksum for Md5 {
+    fn header_name(&self) -> HeaderName {
+        MD5_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        use md5::Digest;
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.finalize();
+        HeaderValue::from_str(&base64::encode(&hash[..]))
+            .expect("will always produce a valid header value from an MD5 checksum")
+    }
```

Comment:
this duplication makes me wonder if we got this API right

Reply:
I can actually replace it with a default impl, I'll do that